src/basicfuns.jl

```diff
@@ -152,9 +152,38 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref).
+
+See:
+ * Martin Maechler (2012) [“Accurately Computing log(1 − exp(− |a|))”](http://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf)
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
-log1pexp(x::Float32) = x < 9.0f0 ? log1p(exp(x)) : x < 16.0f0 ? x + exp(-x) : oftype(exp(-x), x)
+function log1pexp(x::Real)
+    x0, x1, x2 = _log1pexp_thresholds(x)
+    if x ≤ x0
+        return exp(x)
+    elseif x ≤ x1
+        return log1p(exp(x))
+    elseif x ≤ x2
+        return x + exp(-x)
+    else
+        return oftype(x0, x)
+    end
+end
+
+#=
+returns a thresholds x0, x1, x2 such that:
+    * log1pexp(x) ≈ exp(x) for x ≤ x0
+    * log1pexp(x) ≈ log1p(exp(x)) for x0 < x ≤ x1
+    * log1pexp(x) ≈ x + exp(-x) for x1 < x ≤ x2
+    * log1pexp(x) ≈ x for x > x2
+=#
+@inline @generated function _log1pexp_thresholds(::T) where {T<:Real}
+    F = float(T)
+    ϵ = big(eps(F))
+    x0 = log(ϵ / 2)
+    x1 = -log(ϵ) / 2
+    x2 = -log(ϵ/2) - log(-log(ϵ/2)) - log(-log(ϵ/2)) / log(ϵ/2) # root of e^-x == x * ϵ/2
+    return F.((x0, x1, x2))
```

Comment:
With the suggestion above `F` is not needed. Additionally, always good to avoid unnecessary broadcasting operations:
```suggestion
    ϵ = big(eps(T))
    x0 = log(ϵ / 2)
    x1 = -log(ϵ) / 2
    x2 = -log(ϵ/2) - log(-log(ϵ/2)) - log(-log(ϵ/2)) / log(ϵ/2) # root of e^-x == x * ϵ/2
    return (T(x0), T(x1), T(x2))
```

Reply:
Thanks I added the suggested optimizations on the computation of the last threshold. I prefer to keep x0, x1, x2 names though since that's what's used in the paper.